modules/core/src/edu/kit/iti/algover/util/ScriptASTUtil.java

```diff
@@ -0,0 +1,204 @@
+import java.util.logging.Logger;
+
+/**
+ * Utility class for Creating new ScriptASTs. Used for inserting a new Statement
+ * into a given ScriptAST.Script
+ *
+ * @author Valentin Springsklee
+ *
+ */
+public class ScriptASTUtil {
+
+    private ScriptASTUtil() {
+        throw new Error();
+    }
+
+    public static Script insertStatementAfter(Script script, Statement newStatement, ScriptAST referenceASTElem) {
+        // Maybe implement InsertAfterStatementVisitor class, extends
+        // DefaultScriptASTVisitor<Triple<Statement, ScriptAST, ScriptAST>, ScriptAST, SomeException>
+        // pass new Triple<>(newStatement, referenceASTElem, parentAST) a arg
+        ScriptAST updated = script.accept(new UnsealedCopyVisitor() {
+
+            protected void visitStatements(ScriptAST.StatementList old,
+                                           ScriptAST.StatementList newList,
+                                           Statement newStatement,
+                                           ScriptAST referenceASTElem) {
+                for (Statement statement : old.getStatements()) {
+                    if (statement == referenceASTElem) {
+                        newList.addStatement(newStatement);
+                        newStatement.setParent(newList);
+                    }
+                    newList.addStatement((Statement) statement.accept(this, newList));
+                }
+
+                if (old == referenceASTElem) {
+                    newList.addStatement(newStatement);
+                    newStatement.setParent(newList);
+                }
+            }
+
+            @Override
+            public Script visitScript(Script script, ScriptAST arg) throws NoExceptions {
+                Script ret = new Script();
+                visitStatements(script, ret, newStatement, referenceASTElem);
+                ret.setParent(arg);
+
+                return ret;
+            }
+
+            @Override
+            public Case visitCase(Case aCase, ScriptAST arg) throws NoExceptions {
+                Case ret = new Case();
+                ret.setLabel(aCase.getLabel());
+
+                visitStatements(aCase, ret, newStatement, referenceASTElem);
+
+                ret.setProofNode(aCase.getProofNode());
+                ret.setParent(arg);
+
+                return ret;
+            }
+
+
+        }, null);
+
+
+        return (Script) updated;
+    }
+
+    public static Script removeStatementFromScript (Script script, ScriptAST toBeRemoved) {
+        // Also consider separate class as described in insertAfterStatement
+        ScriptAST updated = script.accept(new UnsealedCopyVisitor() {
+
+            protected void visitStatements(ScriptAST.StatementList old,
+                                           ScriptAST.StatementList newList,
+                                           ScriptAST toBeRemoved) {
+                for (Statement statement : old.getStatements()) {
+                    if (statement != toBeRemoved) {
```

Comment:
It works like it is. Because the exact same object(reference) is considered while traversing the script. I think, equals would have to be imlemented for ScriptAST (and maybe its inheriting classes).